SCRUM-253: /users API endpoints added

aas1984 opened the pull request.

aas1984 committed:
feat: added validation package for api endpoints

aas1984 committed:
feat: fleshed out users/ POST

aas1984 committed:
feat: added query parameters to /get api

aas1984 committed:
feat: improved endpoint for updating user

aas1984 committed:
feat: added exception for invalid user id

aas1984 committed:
feat: rebased onto development

aas1984 committed:
test: user.service at 100% coverage

aas1984 committed:
feat: rebased onto development

Jsd5582 commented on src/user/__tests__/user.service.spec.ts:
```diff
@@ -218,4 +220,298 @@ describe('UserService', () => {
+    it('should use displayName when username is specified', async () => {
+      const dto = new GetUsersDto();
+      Object.assign(dto, {
+        username: 'John E. Test',
```
Are we allowing a username to have whitespace? Most apps require a username to have no whitespace

notjoncruz commented on src/user/dto/get-users.dto.ts:
```diff
@@ -0,0 +1,40 @@
+
+export class GetUsersDto {
+  @IsOptional()
+  @ApiPropertyOptional()
```
lets just use `@ApiProperty`

aas1984 replied:
If I don't mark this as `ApiPropertyOptional()` then Swagger complains when I try to manually hit the endpoints from browser.

notjoncruz commented on src/user/dto/get-users.dto.ts:
```diff
@@ -0,0 +1,40 @@
+import { IsBoolean, IsOptional, IsPositive, Max, Min } from 'class-validator';
+import { UserRole } from '../entities/user.entity';
+
+export class GetUsersDto {
```
review the existing use of pagination. we _shouldnt_ need a new dto to support pagination for every entity

notjoncruz commented on src/user/user.controller.ts:
```diff
@@ -20,13 +22,15 @@ import {
+import { GetUsersDto } from './dto/get-users.dto';
+import { UpdateUserDto } from './dto/update-user.dto';
 
 @ApiTags('users')
```
assuming youre gonna use a user decorator for auth

aas1984 replied:
Added base guards to the controller so that it can't be accessed without being logged in. I'll add the custom decorator to `PATCH /:id` to add the extra logic once that PR is merged.

notjoncruz commented on src/user/user.service.ts:
```diff
@@ -102,4 +111,97 @@ export class UserService {
+
+  /**
+   * Returns all users from the database that match the search critera
+   * @param query - The search criteria
```
leverage pagiantion service here

notjoncruz commented on src/user/user.service.ts:
```diff
@@ -102,4 +111,97 @@ export class UserService {
+   * @param id - The id of the user to search for
+   * @returns The user, if they exist, otherwise a BadRequestException
+   */
+  async findOne(id: number) {
```
lets do something like this. in the case that some other module needs to find a user, they can specify other options

https://github.com/The-Mobility-Independence-Foundation/backend/blob/84d860383d4fc10b127ec858ed450bac1980ff4c/src/conversations/conversations.service.ts#L33

aas1984 replied:
Done.

notjoncruz commented on src/user/user.service.ts:
```diff
@@ -102,4 +111,97 @@ export class UserService {
+      throw new BadRequestException();
+    }
+
+    if (dto.firstName) {
```
can we do this instead?
```
Object.assign(...)
```

aas1984 replied:
Done, also changed in a different function.

notjoncruz reviewed (dismissed):
some comments

Jsd5582 commented on src/user/dto/update-user.dto.ts:
```diff
@@ -0,0 +1,54 @@
+  @ApiPropertyOptional()
+  @Min(0, { message: 'Rating must be greater than 0.0' })
+  @Max(5, { message: 'Rating must be less than 5.0' })
+  rating?: number;
```
Get rid of this shit

aas1984 replied:
Done.

aas1984 committed:
refactor: replaced findOne with findById

aas1984 committed:
refactor: changed conditionals to Object.assign()

aas1984 committed:
feat: added guards to the rest of /users

aas1984 committed:
feat: handled comments from PR

Replaced BadRequestExceptions with NotFoundExceptions when appropriate.
Removed Rating from the dtos, service, and tests (this feature is no longer MVP)

aas1984 committed:
feat: rebased onto development

aas1984 committed:
refactor: /users GET now uses cursor pagination
